### ISSUE TITLE
Support a blocklisting of excluded artifact versions

### ISSUE THREAD

**Issue by spyrkob** [wildfly-extras/wildfly-channel#95]
## Overview

When using an open channel, a situation may arise where certain artifacts are promoted to the channel and later discovered to be invalid. As it’s impossible to remove artifacts already deployed into a repository, a mechanism for blocklisting artifacts is needed.

The blocklist should be separate from the channel definition itself so that modifications to the list don’t require changes to the channel definition.

The proposal is to create a ‘blocklist.yaml’ artifact associated with a channel. It will contain a list of streams with excluded versions. When a channel resolves an artifact version, it will consult the latest version of the blocklist. Any versions listed in the blocklist are removed from the stream’s resolved versions and the latest remaining version is treated as the most recent stream version.

When new versions are added to the blocklist, a new version of the blocklist artifact will be promoted into the channel.

## Format
The blocklist file contains a list of stream GAVs with multiple versions:
```
---
exclusions:
  - groupId: io.undertow
    artifactId: undertow-core
    versions: [2.2.18.Final, 2.2.17.Final]
…
```
The artifactId can use a wildcard syntax to refer to all the artifacts with the same groupId
```
---
exclusions:
  - groupId: io.undertow
    artifactId: “*”
    versions: [2.2.18.Final, 2.2.17.Final]
```

## Resolving channel blocklist
A channel can reference blocklist via following header:
```
blocklist:
  groupId: org.channel.foo
  artifactId: blocklist
  versionPattern: “.*”
```

The blocklist is resolved when a channel is initialized. If no blocklist artifact can be resolved with supplied GAVs, the channel treats it as an empty blocklist.
A blocklist applies only to the channel that defined it, not its required channels.

## Resolving artifact version
During artifact version resolution, a stream matching artifact GA is located in the channel. If the stream uses concrete versions, that version of the artifact is resolved and returned to the user.
If the stream uses versionPattern, the blocklist is checked for excluded versions. The excluded versions are removed from the set of available artifact versions before the latest remaining version matching the stream’s pattern is used to resolve the artifact.

If the blocklist excludes all available artifact versions, `UnresolvedMavenArtifactException` is thrown.

The blocklist is ignored when using `resolveDirectMavenArtifact` method.

**Comment by jmesnil** [comment]
I'm not convinced by this RFE. 

A channel is a curated list of components. 
A blocklist is another way to curate by saying I want everything except this and that versions. So now, you are no longer curating the channel but the blocklist :) 

I can see reasons to excludes some components but I'm not sure that adding a new separate artifact for this is the way to go.

* First, using fixed `version` bypasses this whole issue (with more effort to curate the channel).
* Second, using regexp might allow to remove some version (but the resulting regular expression might be quite complex)

If we see a need to curate the channel by dropping some versions, I'd prefer we incorporate that into the channel itself with something like:

```
streams:
  - groupId: org.example
     artifactId: foo
     versionPattern: ".*.Final"
     excludes:
       - 2.2.17.Final
       - 2.2.18.Final
```

wdyt?

**Comment by spyrkob** [comment]
The idea was to allow the channel.yaml of an open channels to be a stateless, static file. The blocklist would become a part of the channel itself like any other artifact.

@wolfc - you had some reservations about the excludes being part of a stream definition

**Comment by spyrkob** [comment]
Fixed by #109
